### DISCUSSION TITLE
Aligning code to paper

### DISCUSSION THREAD

**Discussion by ssardina** [idrave/FOND-ASP#35] [Q&A] [answered]
Going through the paper on correctness and commented text that we may use later (on examples regarding incompleteness on the 2 planers), I clean-up the various planners a lot and made the pretty printing one way more powerful and clean. 

I aligned the code to the paper much more, even if in style, and is now almost identical. Except for one small detail that I am puzzled when reasoning about `blocked/2` for deriving `fair/1`. I want to document this now as I will forget soon and I did a lot of work behind it (not yet fully successful!)

First, the code uses `blocked/2` to derive blocked in any direction between the two nodes, and then `blocked2/2` to capture one-way from `S1` to `S2` blocking. The first 3 rules are all OK and the same:

![image](https://user-images.githubusercontent.com/2368384/156083619-7cdf0c7a-e7eb-4de8-86bc-fe5a1a6a1517.png)

Now, the code in the original codebase is completed with the next rule for `blocked2/2` and the def for `fair/1` as follows **OPTION 1**:

![image](https://user-images.githubusercontent.com/2368384/156084577-a12f48ba-073b-43b7-a846-a0e24375c7b7.png)

The `fair/1` makes total sense and uses `blocked/2` which can reason in any direction. The last `blocked2/2` is more subtle and treats the case of `blocked2(S,S)` separately. 

Now, the code from the paper is **OPTION 2**:

![image](https://user-images.githubusercontent.com/2368384/156084528-c1e00ffd-c78c-46ab-85e9-93ae7ccae722.png)

Note `blocked2` is the one-way direction blocking predicate (it is called `blocked/2` in paper).

So, the question is why @idrave used option 1 variant in the codebase? There seems to be a reason because they don't yield the same result and OPTION 1 gives the correct answer whereas OPTION 2 seems to me may have issues with self-loops. Very puzzling...

The missmatch is shown in [`example03.lp`](domains/clingo/examples/example03.lp), a very simple example that I did for this:

![image](https://user-images.githubusercontent.com/2368384/156084436-9e41eada-fe39-4844-80cb-bec20f9f03e6.png)

The correct solution is given in the OPTION 1 code-base, but this is not exactly what it is in the paper.

So what happens with OPTION 2 (as per paper)? Is it correct or not? The potential problem seems to be in line 69 of OPTION 2 when there is a self-loop, as to get `blocked2(S1,S2)` we will need to prove `blocked2(S1,S2)`! And here minimal models comes to play...

Now, here is the huge surprise, I ran OPTION 2 on example03 and it gives the right solution, despite my intuition that it should not be able to prove `fair(0)` because cannot prove `blocked2(0,2)` and similar. So how does it end-up proving it? I removed the integrity constraint:

![image](https://user-images.githubusercontent.com/2368384/156085288-66ec80b2-f8b6-4b48-a6e7-5fa575f24abd.png)

This is the ONLY integrity constraint. As expected OPTION 1 still yields 1 model, the solution.

But OPTION 2 yields 2 models!! One is the one I think is intuitively correct, where due to line 69, it cannot prove `blocked2(0,2)` due to the self loop on `0`, and hence no termination (which would violate the integrity constraint, but we left it out). 

However, the second model given by OPTION 2 is, surprisingly, a SUPER SET, of the first model and has all the correct atoms! That is, it now adds `blocked2(0,2)` and many more, including correct  `terminates/2` for everyone, etc. 

What is happening is that when the integrity constraint is added, then the first model is cut down, and the second remains.

Clearly, there is something very wrong that I am not understanding, but:

1. Why OPTION 2 gives 2 models when the integrity constraint is dropped? And why 1 is a superset, how come if they are stable models?
2. More seriously, is it true that OPTION 2 would not work well under self-loops. This is more suspecious as @idrave used an alternative in the code base.

OK I finish documenting this as detailed as I can to be able to resume it. There is a lot of good code left to test this (and many other things) in the new version of the planners.

@bonetblai , do you have time to go over this with me? I suspect that I cannot see anymore the glitch, but someone from "outside" will see the issue right away... :wink:

**Comment by idrave** [comment]
I have checked your observations about the two options.

What happens with OPTION 2 is tricky to see, but it has an explanation and is related to the use of negation in Clingo, particularly when it appears on the right side of the conditional literal in line 70 (`blocked2(T,S) : ... not blocked2(S,T).`).

If we consider a simplified version of what the rule for fair(0) looks like given the graph and policy above:

```prolog
fair(0) :- blocked2(2,0) : not blocked2(0,2)
```

As you noted, given that we have no way to infer `blocked(0,2)`, then we can assume `not blocked2(0,2)` to be true. Then we will need `blocked2(2,0)` to hold in order to prove `fair(0)`, but because of the self-loop on state `0`, `blocked2(2,0)`  on lines `68-69` depends on `blocked2(2,0)`, and the program is stuck.

However, as you saw there is another stable model that is a superset of this previous one. The issue is that the stable models of a Clingo program are _not guaranteed to be minimal when there is negation_ in the program. In particular, the stable models `M` of a program `P` with negation are the minimal models of the Clingo program resulting from replacing all negative literals in `M` by their truth-value given model `M` (the reduct of `P` relative to `M`) [1], they are not the minimal models of `P` itself. 

Then, what happens on `fair(0) :- blocked2(2,0) : not blocked2(0,2)` is that if we assume `blocked2(0,2)`, then we can infer `fair(0)` and subsequently `terminate(0)` and `blocked(0,2)`. Since we could infer `blocked(0,2)`, then the model is still minimal wrt the reduct of the program.

This might be counterintuitive to the idea that in Clingo we only add the head of a rule to the model when we can prove the body of the rule. However, when there is negation it is often not that simple, especially when there are cyclic dependencies as in our program. **This is why in the code base (OPTION 1) I tried to structure the code** in such a way that the right side of conditions only used positive literals, even when it implied needing to define two terms `blocked/2` and `blocked2/2`.

Regarding self-loops, in OPTION 1 I added condition `S1 != T` in the rule for `blocked2/2` in order to avoid `blocked2(S1,S2)` depending from `blocked2(S1,S2)` when it should not get stuck in there.

In the case of OPTION 2, it overcomes the issue of the self-loop here, but it seems to me now that those rules might accept policies that are not correct in certain problems, due to the issues with the definition of stable models I mention above. I was thinking about it and [example03_2.lp](https://github.com/idrave/FOND-ASP/blob/master/domains/clingo/examples/example03_2.lp) should be unsat, but OPTION2 gives a solution, while OPTION 1 returns unsat.

[1] See https://link.springer.com/content/pdf/10.1007%2F978-3-030-24658-7.pdf section 5

**Reply by ssardina** [reply]
Hola @idrave .

Spot on, you are totally right; I keep forgetting the stable models for non-positive programs may not be minimal (BTW, it is not an issue of Clingo system itself, it is the def of stable model), as it is the case with the simple program `p <- not not p` which has to answer sets (stable models): the empty set `{}` and `{p}` and obviously the second is a super set of the first one, so not minimal per se (minimal on the reduct yes). This "acyclic" property is quite convoluted in our program because it involves more than 1 predicate and it involves conditions.

But your reasoning kind of confirms my suspicious that OPTION 2 may not work always even if it works on the self-loop. I saw you pushed a variant of my example 3, great, I am checking that one carefully today to see if we converge. :-)

Thanks for following up. Hopefully we'll find a stable answer soon :+1:

**Reply by ssardina** [reply]
I ran the `example03_2` (good you used the digraph encoding, thanks!):

![image](https://user-images.githubusercontent.com/2368384/156707798-7998047f-db45-4655-aa1f-b38763afb9b1.png)

Clearly the loop `0 -a-> 2` is a trap and hence there could be no solution.

But OPTION 2 (the paper) yields one solution! And it just gives 1 model (even without the integrity constraint)

We are getting somewhere here. So OPTION 1 is NOT equivalent to OPTION 2

**Reply by ssardina** [reply]
OK @idrave , let's see if we are on the same page. Clearly [example03_2.lp](https://github.com/idrave/FOND-ASP/blob/master/domains/clingo/examples/example03_2.lp)  should be UNSAT, but OPTION 2 would yield SAT and that is wrong. The reason is that OPTION 2 will infer `fair(0)` wrongly because:

1. we have `fair(0) :- blocked(2,0) : not blocked(0,2)`
2. because of the `not`, we will make an assumption and then check the minimal models of the reduct. So one assumption could be that `blocked(0,2)` holds true
3. Under that assumption `fair(0) :- true` and hence `fair(0)` is derived.
4. Then `terminates(0)` follows because we have `fair(0)` and `terminates(1)` (1 is fair and has a goal successor)
5. Then `terminates(2)` is derived bc the only successor is `0`, which is now terminating.
6. Done, all states have been marked as terminating.

So, the error here is how we check the notion that _"a loop with a B-action has a terminating state". We are using the trick with the `not` but that seems to NOT be correct trick. _Agree?_ This is very subtle bc it requires a deep understanding of AS under negation.

**Comment by ssardina** [comment, marked as answer]
This was all solved in the paper repo. :-)
